Fix cart voucher values in cart summary and checkout summary

arouiadib opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x 
| Description?  | This PR fixes the issues of the amounts of promo codes in cart summary and checkout summary.
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? |  Fixes #16842 
| How to test?  | Check the amount of reduction of each cart voucher, now it depends of the actions of the cart voucher and is not equal to all products total.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/17502)
<!-- Reviewable:end -->

arouiadib committed:
fixing the calculation of amount of a cart rule when this cart rule has percentage reduction

the amount depends on the type of percentage reduction (cheapest product, specific product, selection of products or whole order). This processing already exists in `getContextualValue`in CartRule.php.

MatShir commented:
@arouiadib: "the PR I created is a "white-flag" PR for me, because I have spend hours trying to create a cleaner and better solution to no avail. My one-line PR uses an existing method getContextualValue in CartRule.php. Does it go against migration principles? Better not to use methods from that class? Am I right? Additionally, the processing in that mothod almost exists in CartRuleCalculator also, but not easy to re-use ( as I see it addresses another purpose IMO, and re-factoring should be done carefully).

Another possible solution is use quantity from product list in Presenter, but it would be a second duplication for that processing, and maybe is not the role of cartPresenter?!

I think for a cleaner solution is to make a re-factoring in CartRuleCalculator. But more experienced people from the community will suggest me the right direction."

@PrestaShop/prestashop-core-developers What do you think ?

marionf commented:
For @PrestaShop/prestashop-qa 
Could you verify if this PR brings no regressions with a free shipping cart rule (https://github.com/PrestaShop/PrestaShop/issues/9702) ?

arouiadib committed:
Rename variable to reflect the current need

arouiadib committed:
Enhanced the solution to incorrect promo value

No need to use `getContextualValue()` method from `CartRule` class, directly get reduction amount from $cartVoucher.

arouiadib committed:
Fix promo value in cart summary and cart checkout summary

Rewrote the function so that the shipping reduction is not added to total reduction when cart rule has free shipping.

arouiadib committed:
Fix Coding standards issues

arouiadib committed:
Fix More coding standard issue

arouiadib committed:
If multiple vouchers with free shipping, only display one

arouiadib committed:
Coding style

arouiadib committed:
Fixed merge conflict

PierreRambaud commented on src/Adapter/Presenter/Cart/CartPresenter.php:
```diff
@@ -640,6 +637,16 @@ private function cartVoucherHasAmountReduction($cartVoucher)
+     */
+    private function cartVoucherHasGiftProductReduction($cartVoucher)
+    {
+        return array_key_exists('gift_product', $cartVoucher) && $cartVoucher['gift_product'];
```
```suggestion
        return !empty($cartVoucher['gift_product']);
```

Is it really needed to add a method only for this check?

arouiadib replied:
Not really needed, but just to follow suit with current code.

arouiadib replied:
I see what you mean now. Well, the only reason is to make the `if` statement more self-explanatory  ( the naming `cartVoucherHasGiftProductReduction` inside `if` emphasizes the fact that we need the 3 conditions that have "similar" weight). But you know, I am still learning the codebase so I can't tell which one is more suitable until you directly tell me ;-)

PierreRambaud replied:
In fact, this is the same thing.
```
!empty($cartVoucher['gift_product']); // => Not false, 0, '', null, []
array_key_exists('gift_product', $cartVoucher) && $cartVoucher['gift_product'] // => Check if the key exists in the table and is not null, false, 0, '', etc
```

This is the same check, but it's faster due to only one call

arouiadib replied:
Thank you very much :). 
If I understand well (because I really have a slow brain these days), we can we go further by adding [one another commit](https://github.com/PrestaShop/PrestaShop/pull/17502/commits/099a863b16a8a3405af20d28269b483ccb015cb0).

arouiadib committed:
PierreRambaud suggestion

Co-Authored-By: GoT <PierreRambaud@users.noreply.github.com>

Progi1984 commented:
@arouiadib Hi Could you fix the PHPCSFixer ? Thanks

arouiadib commented:
@Progi1984 , I tried today to no avail. I will retry ASAP. Thank you very much for your time.

arouiadib committed:
PHP-CS

arouiadib committed:
PHP-CS

arouiadib committed:
Merge branch 'Fix-issue-16842' of https://github.com/arouiadib/PrestaShop into Fix-issue-16842

arouiadib committed:
Keep initialization of variable

arouiadib committed:
Merge branch 'Fix-issue-16842' of https://github.com/arouiadib/PrestaShop into Fix-issue-16842

arouiadib committed:
Rebased

arouiadib commented:
Hi @Progi1984 , I finally managed to fix code styles issue,  messed up a bit with the branch :) 
Thank you very much for your time.

arouiadib committed:
Refactor cartVoucherHasGiftProductReduction()

SD1982 commented:
hi @arouiadib  Can you please resolve the conflict ?

Progi1984 commented:
Hello @arouiadib, thanks for your contribution, would you mind to rebase your pull request? Thanks

arouiadib commented:
Hello @SD1982 @Progi1984 , I think this PR is no longer taking the direction that the code source is taking. I will be rethinking things to bring a more appropriate change. Thank you very much guys for all your time.

arouiadib committed:
resolve merge conflicts

arouiadib committed:
Merge branch 'Fix-issue-16842' of https://github.com/arouiadib/PrestaShop into Fix-issue-16842

arouiadib committed:
Move the fix of @Progi1984 for #18065 from CartPresenter.php to CartRule.php

arouiadib committed:
Resolve PHP CS Fixer issue

matthieu-rolland commented:
@arouiadib you're almost done, one last conflict to fix ;)

arouiadib committed:
Merge branch '1.7.7.x' into Fix-issue-16842

arouiadib committed:
PHP CS Fixer

jolelievre commented:
Hi @PrestaShop/prestashop-qa 

In addition to what @marionf said about checking no regression regarding the issue #9702 about free shipping cart rule

You should also check this issue https://github.com/PrestaShop/PrestaShop/issues/18065 since lots of the code that fixed this in https://github.com/PrestaShop/PrestaShop/pull/18144 has been replaced

So it would seem all this calculation was useless because it is present in `$cartVoucher` which would be great! @Progi1984 since you approved this PR and coded the previous one i guess it's ok but since we don't have much automated test on this code better safe than sorry ^^

SD1982 commented:
LGTM I have also checked for #9702 and #18065 as suggested by @jolelievre and i haven't notified any unwanted behaviour or regression ! Thanks @arouiadib !

Progi1984 commented:
Thanks @arouiadib

arouiadib commented:
Thank you all guys.

ComonSoft commented:
Hi, this fix the issue #16842 but the reduction summury is still false in cart summury see screenshot
![image](https://user-images.githubusercontent.com/13507563/94287172-4352ed80-ff56-11ea-91f4-48fb243ddc2a.png)

MatShir commented:
Hi @ComonSoft,
Could you create a new issue with the detailed steps to helps us reproduce it?
We will be happy to test and check it out.

ComonSoft commented:
@MatShir this is the same issue as describre in ticket #19090 The solution proposed here only solve the amount of each voucher in cart summury but not the sum of all voucher.

MatShir commented:
Okay thanks for your feedbacks @ComonSoft, the bug is placed on the Roadmap 1.7.8 ;-)
